feat: `create-release`

marek-saji opened the pull request:
# Why?

Closes https://verkstedt.atlassian.net/browse/VIP-87

# What?

New manually triggered workflow that uses https://www.npmjs.com/package/semantic-release under the hood, but in a non–standard way:

- workflow input dictates release type (usually commit messages dictate this)
- GitHub API is used to generate release notes, including PR titles and authors (usually commit messages are used for this)
- we modify files (`package.json`, `CHANGELOG.md` and `version.txt` (if it exists)) and push them. This is not recommended by semantic-release, but a common practice. A custom GitHub token is needed for this.

We also do more standard things:

- create and push git version tag
- create GitHub release

# Review notes

> [!IMPORTANT]
> Remember this is a public repo. Do not discuss internal projects.

## Testing

There’s a private test repo set up at https://github.com/verkstedt/actions--test it sends notifications to `#infrastructure-create-release-pr-action` slack channel.

# TODO before this is merged

- [x] Change `@feat/create-release` to `@v1`

# TODO after this is merged

- [x] Add workflow template to `verkstedt/.github` repo. Will look something like https://github.com/verkstedt/actions--test/blob/main/.github/workflows/create-release.yaml
- [x] Add to `@verkstedt/setup-repo`

# Next steps

Whole `semantic-release` setup could probably be extracted to separate npm package that it can be called with `npx`, instead of having to be installed, but let’s do that after we prove this approach useful at all.

marek-saji committed:
create-release-pr: Pretty white space in CHANGELOG

marek-saji committed:
feat(create-release): Use semantic-release

marek-saji committed:
chore: Kebab case for inputs

marek-saji committed:
create-release-pr

marek-saji committed:
create-release: new vision

marek-saji committed:
create-release: Remove working-dir input, until it’s needed

marek-saji committed:
create-release: dry-run

marek-saji committed:
create-release: Triggering GH user in git commit

marek-saji committed:
create-release: Use ::error::

marek-saji committed:
fix(create-release): Piping release notes to step summary

marek-saji committed:
create-release: Notify about success

marek-saji committed:
create-release: No notify on dry-run

marek-saji committed:
create-release: Use GitHub app token

marek-saji commented on .github/workflows/create-release.yaml:
```diff
@@ -0,0 +1,204 @@
+# Create Release
+#
+# Template to use with this workflow:
+# https://github.com/verkstedt/.github/tree/main/workflow-templates/create-release.yaml
```
Will be added after this PR is merged and look like this: https://github.com/verkstedt/actions--test/blob/main/.github/workflows/test-create-release.yaml (but with `v1` instead of `feat/create-release`).

marek-saji committed:
Remove TODOs

marek-saji committed:
doc(create-release)

marek-saji commented on .github/workflows/create-release.yaml:
```diff
@@ -0,0 +1,204 @@
+      issues: write
+
+    steps:
+      - name: Verify inputs
+        run: |
+          if ! echo "${{ inputs.release-type }}" | grep -qE '^(major|minor|patch|prerelease)$'
+          then
+            echo "::error::Invalid release-type: ${{ inputs.release-type }}. Must be one of: major, minor, patch, prerelease."
+            exit 64 # EX_USAGE
+          fi
+
+          if [ "${{ inputs.release-type }}" = "prerelease" ]
+          then
+            echo "::error::Prerelease not supported (yet)."
+            exit 64 # EX_USAGE
+          fi
```
Later.

marek-saji commented on .github/workflows/create-release.yaml:
```diff
@@ -0,0 +1,204 @@
+          user_id=$(gh api "/users/${{ steps.app-token.outputs.app-slug }}[bot]" --jq .id)
+          echo "user-id=$user_id" >> "$GITHUB_OUTPUT"
+
+      - name: Setup
+        id: setup
+        uses: verkstedt/actions/setup@v1
+        with:
+          token: ${{ steps.app-token.outputs.token }}
+          github-npm-registry-personal-access-token: ${{ secrets.GH_NPM_REGISTRY_PERSONAL_ACCESS_TOKEN }}
+
```
Not true.

marek-saji committed:
DEBUG(create-release): Use setup@feat/create-release

marek-saji commented:
> > > Now that this section documents our default branch as `v1`, the “Assumptions” list still stating “Main branch is called `main`.” is contradictory. Please update that earlier bullet to keep the README self-consistent.

@coderabbitai This is an assumption about the project using these actions, not the actions repo itself.

marek-saji commented on .github/workflows/create-release.yaml:
```diff
@@ -0,0 +1,209 @@
+            exit 64 # EX_USAGE
+          fi
+
+          if [ "${{ inputs.release-type }}" = "prerelease" ]
+          then
+            echo "::error::Prerelease not supported (yet)."
+            exit 64 # EX_USAGE
+          fi
```
Will be implemented later.

marek-saji committed:
create-release: Note about $s

marek-saji committed:
create-release: Skip previous_tag_name, if there’s none

marek-saji committed:
create-release: Custom success PR comment

marek-saji committed:
fixup! create-release: Skip previous_tag_name, if there’s none

marek-saji commented on .github/workflows/create-release.yaml:
```diff
@@ -0,0 +1,239 @@
+          # inputs.* directly to make it easier to copy it to separate
+          # file for testing
+          GITHUB_TOKEN: ${{ steps.app-token.outputs.token }}
+          PACKAGE_MANAGER: ${{ steps.setup.outputs.package-manager }}
+          RELEASE_TYPE: ${{ inputs.release-type }}
+        run: |
```
both are recognised

marek-saji committed:
fixup! create-release: Note about $s

marek-saji commented on .github/workflows/create-release.yaml:
```diff
@@ -0,0 +1,209 @@
+        uses: verkstedt/actions/setup@feat/create-release
+        with:
+          token: ${{ steps.app-token.outputs.token }}
+          github-npm-registry-personal-access-token: ${{ secrets.GH_NPM_REGISTRY_PERSONAL_ACCESS_TOKEN }}
```
It’s only temporary here. Will revert before merging, like stated in PR description.

marek-saji committed:
fixup! create-release: Custom success PR comment

marek-saji committed:
create-release: unindent shell commands

marek-saji committed:
fixup! create-release: Custom success PR comment

marek-saji committed:
fixup! create-release: Use GitHub app token

marek-saji commented on setup/action.yaml:
```diff
@@ -10,6 +10,10 @@ inputs:
     description: 'Directory in which to run things in.'
     required: false
     default: '.'
+  token:
+    description: 'GitHub token. Defaults to `github.token`.'
+    required: false
+    default: ${{ github.token }}
```
Tested. Works fine.

marek-saji committed:
create-release: Improve setup instructions

marek-saji committed:
create-release: docs

smelchior commented on .github/workflows/create-release.yaml:
```diff
@@ -0,0 +1,264 @@
+          GIT_AUTHOR_EMAIL='${{ steps.app-token-data.outputs.user-id }}+${{ steps.app-token.outputs.app-slug }}[bot]@users.noreply.github.com'
+          export GIT_AUTHOR_NAME GIT_AUTHOR_EMAIL
+
+          # Not really semantically true, because semanit-release is
```
nitpick: typo for semanit-release

smelchior commented on .github/workflows/create-release.yaml:
```diff
@@ -0,0 +1,264 @@
+        run: |
+          echo "::group::Prepare semantic-release configuration"
+          release_config_mjs="$( mktemp --suffix=.mjs )"
+          cat > "$release_config_mjs" << 'RELEASE_CONFIG_MJS_EOF'
```
I think having this embedded in here is a bit complicated and hard to read. Could we put it in an extra file and reference it via ${{ github.action_path }}? The required variables could still be pulled from the env var on execution?

marek-saji replied:
Oh, nice. I didn’t know about `github.action_path`. That makes much more sense.

Like stated in the comment, my plan was to eventually make it a separate npm package or something, but maybe that won’t be necessary.

marek-saji replied:
sorry for spamming :facepalm: Noticed only now, because GH notifications are laggy.

smelchior commented on .github/workflows/create-release.yaml:
```diff
@@ -0,0 +1,264 @@
+                  .filter(p => ! /^(\\.\\/|\\.\\.\\/|\/)/.test(p))
+                  .join('\\n')
+              );
+            " "$release_config_mjs" |
```
as we know the plugins from the file, and it is static in here anyway, can we maintain this as a list? This seems rather complex for the task.

marek-saji replied:
How would you feel if I’d keep it dynamic here, but greatly simplify? Commit will follow.

smelchior replied:
that is a lot nicer now, thanks :)

marek-saji committed:
fixup! create-release: Triggering GH user in git commit

Resolves https://github.com/verkstedt/actions/pull/25#discussion_r2452355090

marek-saji committed:
create-release: Move config to separate file

Resolves https://github.com/verkstedt/actions/pull/25#discussion_r2452370872

marek-saji committed:
Merge remote-tracking branch 'origin/v1' into feat/create-release

marek-saji committed:
create-release: New version and notes in notification

marek-saji commented on .github/workflows/create-release.yaml:
```diff
@@ -0,0 +1,208 @@
+            git restore package-lock.json || :
+            git restore yarn.lock || :
+          }
+          install-plugin-pkgs
```
I, myself prefer to shy away from bashisms in my shell scripts.

marek-saji committed:
fixup! create-release: New version and notes in notification

Resolves https://github.com/verkstedt/actions/pull/25#discussion_r2455492486

marek-saji commented on package.json:
```diff
@@ -20,8 +20,12 @@
     "npm-run-all": "^4.1.5",
-    "prettier": "^3.2.5"
+    "prettier": "^3.2.5",
+    "semantic-release": "^25.0.1"
```
You are hallucinating.
These vulns are for lower versions.

marek-saji committed:
fixup! create-release: New version and notes in notification

Resolves https://github.com/verkstedt/actions/pull/25#discussion_r2455492468

marek-saji committed:
create-release: Quote shell env vars

Resolves https://github.com/verkstedt/actions/pull/25#discussion_r2455561392, https://github.com/verkstedt/actions/pull/25#discussion_r2455561411 and https://github.com/verkstedt/actions/pull/25#discussion_r2455561422

marek-saji commented on create-release/semantic-release.config.mjs:
```diff
@@ -0,0 +1,176 @@
+        `),
+      },
+    ],
+
+    // Commit and push changed files
+    [
+      '@semantic-release/git',
+      {
+        assets: [
+          'CHANGELOG.md',
+          'package.json',
+          'package-lock.json',
+          'version.txt',
+        ],
+        message: unindent(`
+          chore(release): \${nextRelease.version}
+
+          \${nextRelease.notes}
+        `),
+      },
+    ],
```
this is not true. `assets` array for `@semantic-release/git` CAN contain non–existing files and they will just be ignored. I tested it out.

smelchior commented on setup/action.yaml:
```diff
@@ -10,6 +10,10 @@ inputs:
     required: false
     default: '.'
+  token:
+    description: 'GitHub token. Defaults to `github.token`.'
```
The Token is used for cloning the repo? I am not quite sure if that is necessary as the checkout will be always for the repo it runs in? So we do not really need it?

marek-saji replied:
That’s a good point.
I’m 99.9% sure removing this will not break anything (https://github.com/search?q=path%3A.github%2Fworkflows%2F+%22verkstedt%2Factions%2Fsetup%22&type=code), but just to be on a safe side, I’ll create separate issue and remove it once all other open PRs are merged.

https://verkstedt.atlassian.net/browse/VIP-92

smelchior replied:
As discussed, this is required for pushing to protected branches, so we want to keep it, but update the description

smelchior commented on setup/action.yaml:
```diff
@@ -10,6 +10,10 @@ inputs:
+    required: false
+    default: ${{ github.token }}
   github-npm-registry-personal-access-token:
     description: 'GitHub personal access token for npm registry authentication. You can create one at https://github.com/settings/tokens/new (scopes: `read:packages`, `repo` and `read:org`). If not set, will use secrets.GITHUB_TOKEN, which is enough for most cases.'
```
I am not sure adding personal tokens here in secrets is a good idea as if people leave things will break at some point. 
https://docs.github.com/en/packages/learn-github-packages/configuring-a-packages-access-control-and-visibility -> There is the option to allow other repos access to the package, i think this should be pushed to be used in these cases?

marek-saji replied:
Ideally yes, but we are using it for repos installed from a different org and it’s difficult to ask ops there to change anything.

smelchior replied:
Extend description to follow this order for the tokens:
* Use package access from https://docs.github.com/en/packages/learn-github-packages/configuring-a-packages-access-control-and-visibility -> no token needed
* USe token of *bot* user (we still need to define which one)

-> Not use a personal token from your own GH account

marek-saji commented:
@coderabbitai pause

marek-saji committed:
setup: Expand token–related descriptions

Resolves https://github.com/verkstedt/actions/pull/25#discussion_r2498287466
Resolves https://github.com/verkstedt/actions/pull/25#discussion_r2498298811

marek-saji committed:
setup: Multi–line descriptions

marek-saji committed:
Typo

marek-saji committed:
create-release: inputs, not vars

marek-saji committed:
create-release: Drop ENV_VARS

smelchior commented on README.md:
```diff
@@ -173,6 +173,13 @@ Template:
+### Create Release
+
+Template:
+<https://github.com/verkstedt/.github/tree/main/workflow-templates/create-release.yaml>
```
The template is currently not present (yet)

marek-saji replied:
See “TODO after this is merged” in the PR body :slightly_smiling_face:

marek-saji committed:
fixup! create-release: New version and notes in notification

marek-saji committed:
feat(notify-status): Include text–only value

For OS notifications and such.

marek-saji committed:
feat(create-release): Graceful when slackifying fails

marek-saji committed:
fixup! feat(notify-status): Include text–only value

marek-saji committed:
fixup! fixup! feat(notify-status): Include text–only value

marek-saji committed:
feat(create-release): Update package.json if exists only

marek-saji committed:
fix(create-release): Drop type:module import assertion

We already run whole node with --input-type=module

marek-saji committed:
fixup! feat(create-release): Update package.json if exists only

marek-saji committed:
feat(create-release): Check other workflows before running

marek-saji committed:
Merge remote-tracking branch 'origin/v1' into feat/create-release

marek-saji committed:
fixup! feat(create-release): Check other workflows before running

marek-saji committed:
fix(create-release): Brute force when installing semantic-release plugins

marek-saji committed:
feat(create-release): Split Release step

marek-saji committed:
Merge remote-tracking branch 'origin/v1' into feat/create-release

marek-saji committed:
chore(create-release): Switch versions from feat branch to v1

marek-saji committed:
chore: prettier -w .
